lib/redis_memo/memoize_query/invalidation.rb

```diff
@@ -123,50 +123,89 @@ def self.rewrite_import_method(model_class, method_name)
         # - default values filled by the database
         # - updates on conflict conditions
         records_to_invalidate =
-          if unique_by
+          if columns_to_update
             RedisMemo::MemoizeQuery::Invalidation.send(
-              :select_by_uniq_index,
+              :select_by_columns,
               records,
-              unique_by,
+              columns_to_update,
             )
           else
             []
           end
 
         result = send(:"#{method_name}_without_redis_memo_invalidation", *args, &blk)
 
-        records_to_invalidate += RedisMemo.without_memo do
-          # Not all databases support "RETURNING", which is useful when
-          # invaldating records after bulk creation
-          model_class.where(model_class.primary_key => result.ids).to_a
+        # Offload the records to invalidate while selecting the next set of
+        # records to invalidate
+        case records_to_invalidate
+        when Array
+          RedisMemo::MemoizeQuery.invalidate(*records_to_invalidate) unless records_to_invalidate.empty?
+
+          RedisMemo::MemoizeQuery.invalidate(*RedisMemo::MemoizeQuery::Invalidation.send(
+            :select_by_id,
+            model_class,
+            # Not all databases support "RETURNING", which is useful when
+            # invaldating records after bulk creation
+            result.ids,
+          ))
+        else
+          RedisMemo::MemoizeQuery.invalidate_all(model_class)
         end
 
-        memos_to_invalidate = records_to_invalidate.map do |record|
-          RedisMemo::MemoizeQuery.to_memos(record)
-        end
-        RedisMemo::Memoizable.invalidate(memos_to_invalidate.flatten)
-
         result
       end
     end
   end
 
-  def self.select_by_uniq_index(records, unique_by)
+  def self.select_by_columns(records, columns_to_update)
     model_class = records.first.class
     or_chain = nil
-
-    records.each do |record|
-      conditions = {}
-      unique_by.each do |column|
-        conditions[column] = record.send(column)
+    columns_to_select = columns_to_update & RedisMemo::MemoizeQuery
```

Comment:
Can we add unit tests / modify existing ones to check the logic we originally missed here?

E.g. if we're importing `Site.import(records, on_duplicate_key_update: [:a, :b])`, but we only memoize the column :a, we shouldn't be querying / invalidating :b

Reply:
Good idea! Will add this test case.